velox/exec/Unnest.cpp

```diff
@@ -146,22 +145,81 @@ RowVectorPtr Unnest::getOutput() {
   return output;
 }
 
-void Unnest::generateRepeatedColumns(
-    vector_size_t start,
+const Unnest::RowRange Unnest::extractRowRange(
     vector_size_t size,
+    vector_size_t& numElements,
+    bool& lastRowPartial) {
+  vector_size_t numInput = 0;
+  std::optional<vector_size_t> lastRowEnd;
+  for (auto row = nextInputRow_; row < size; ++row) {
+    bool isFirstRow = (row == nextInputRow_);
+    vector_size_t remainingSize =
+        isFirstRow ? rawMaxSizes_[row] - firstRowStart_ : rawMaxSizes_[row];
+    ++numInput;
+    if (numElements + remainingSize > maxOutputSize_) {
+      // Single row's output is divided into multiple batches.
+      // Computes the row range to process the first and last row
+      // partially instead of 0 to rawMaxSizes_[row].
+      if (isFirstRow) {
+        lastRowEnd = firstRowStart_ + maxOutputSize_ - numElements;
+      } else {
+        lastRowEnd = maxOutputSize_ - numElements;
+      }
+      lastRowPartial = true;
+      // Process maxOutputSize_ in this getOutput.
+      numElements = maxOutputSize_;
+      break;
+    } else {
+      // Not split this row.
+      numElements += remainingSize;
+    }
+    if (numElements >= maxOutputSize_) {
+      break;
+    }
+  }
+  // The last row is not partial, set it to the maxSize.
+  if (!lastRowEnd.has_value()) {
+    lastRowEnd = rawMaxSizes_[nextInputRow_ + numInput - 1];
+  }
+  return {nextInputRow_, numInput, lastRowEnd.value()};
+};
+
+void Unnest::generateRepeatedColumns(
+    const RowRange& range,
     vector_size_t numElements,
     std::vector<VectorPtr>& outputs) {
   // Create "indices" buffer to repeat rows as many times as there are elements
   // in the array (or map) in unnestDecoded.
   auto repeatedIndices = allocateIndices(numElements, pool());
   auto* rawRepeatedIndices = repeatedIndices->asMutable<vector_size_t>();
   vector_size_t index = 0;
-  for (auto row = start; row < start + size; ++row) {
+  VELOX_DCHECK_GT(range.size, 0);
+  // Record the row number to process.
+  // Process the first row.
+  const auto firstRowEnd =
+      range.size == 1 ? range.lastRowEnd : rawMaxSizes_[range.start];
+  std::fill(
```

Comment:
Should we use std::fill in the loop over middle rows as well?

Reply:
Sure